webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,739 @@
+    option.setAttribute("provider", this._formatProviderName(providerName));
+    return option;
+  },
+
+  _addNotificationMultipleProviders(providersMap, document) {
+    const multiProviderSelect = document.getElementById("multiProviderSelect");
+
+    const providers = [];
+    providersMap.forEach((value, key) => {
+      providers.push({"key": key, "name": value.displayName});
+    });
+
+    // Reset multiProviderSelect by removing any previous
+    // added options
+    while (multiProviderSelect.firstChild) {
+      multiProviderSelect.firstChild.remove();
+    }
+
+    for (const provider of providers) {
+      multiProviderSelect.appendChild(
+        this._addRadioOption(provider.key, provider.name, document));
+    }
+    multiProviderSelect.removeAttribute("hidden");
+  },
+
+  _iconURL(name) {
+    if (this._formatProviderName(name) === "localdownloads") {
+      return "moz-icon://" + CloudDownloadsInternal.defaultDownloadDirIconURL + "?size=16";
+    }
+    return new URL(this.stylesURL).origin + "/skin/" + this._formatProviderName(name) + ".svg";
+  },
+
+  _setMenuItemAttributes(menuItem, providerKey, providerDetail, downloadElement) {
+    menuItem.setAttribute("label", providerDetail.displayName);
+    menuItem.setAttribute("providerKey", providerKey);
+    menuItem.setAttribute("class", "menuitem-iconic");
+    menuItem.setAttribute("image", this._iconURL(providerDetail.displayName));
+    return menuItem;
+  },
+
+  _setMoveDownloadMenuPopUpAttributes(menuItem) {
+    menuItem.removeAttribute("hidden");
+    const menuPopup = menuItem.menupopup;
+    menuPopup.removeAttribute("hidden");
+    return menuPopup.firstChild;
+  },
+
+  _displayMoveToCloudContextMenuItem(downloadElement) {
+    const aPopupMenu = downloadElement.ownerDocument.getElementById("downloadsContextMenu");
+    let menuItem = aPopupMenu.getElementsByAttribute("id", "moveDownload")[0];
+    if (menuItem) {
+      menuItem.setAttribute("hidden", "true");
+
+      // Exit if user has opted to store subsequent downloads in cloud provider folder
+      if (CloudDownloadsInternal.preferredProviderKey) {
+        return;
+      }
+
+      if (CloudDownloadsInternal.checkIfExistingCloudProviderDownloadSettings()) {
+        return;
+      }
+
+      const downloadType = downloadElement.getAttribute("cloudstorage");
+      // Downloaded Item is saved in local downloads folder
+      if (this.providers.size > 1) {
+        let subMenuItem =  this._setMoveDownloadMenuPopUpAttributes(menuItem);
+        this.providers.forEach((value, key) => {
+          if (this._formatProviderName(value.displayName) !== downloadType) {
+            subMenuItem = this._setMenuItemAttributes(subMenuItem, key, value, downloadElement);
+            subMenuItem = subMenuItem.nextSibling;
+          }
+        });
+        if (downloadType != "local") {
+          subMenuItem = this._setMenuItemAttributes(subMenuItem, "local", {displayName: "Local Downloads"}, downloadElement);
```

Comment:
So this works... but kind of by accident. In the sense that there's a fixed number of items in this popup (namely, N for N providers), and either the download is local and then we don't need the local item, or it belongs to one of the cloud providers and then we miss out one of those and reuse it for the 'local' case instead.

But this feels pretty fragile, and in particular I think the more usual thing to do here would be to just create all N+1 items and mark them as hidden depending on where the download is. As it is, if there's ever a mismatch between provider keys, there'll be JS errors showing this context menu because there won't be enough children in the menupopup to complete the loop and run this item.

Reply:
Updated to create N+1 items